tools/distrib/upgrade_boringssl_objc.sh

```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
+# Copyright 2018 gRPC authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# Generate the list of boringssl symbols that need to be shadowed based on the
+# current boringssl submodule. Requires local toolchain to build boringssl.
+
+set -e
+
+cd "$(dirname $0)"
+cd ../../third_party/boringssl
+
+BORINGSSL_COMMIT=$(git rev-parse HEAD)
+BORINGSSL_PREFIX_HEADERS_DIR=src/objective-c/boringssl_prefix_headers
+
+# Do the following in grpc root directory
+cd ../..
+
+docker build tools/dockerfile/grpc_objc/generate_boringssl_prefix_header -t grpc/boringssl_prefix_header
```

Comment:
do we really need docker for this, it seems that generating the prefix header should work just fine on people's workstation?
It look like you need 3 non-trivial scripts upgrade_boringssl_objc.sh, Dockerfile and generate_boringssl_prefix_header.sh to update the prefix header, so I'm not sure this actually simplifies things?

Reply:
I am using Docker for a few reasons:
* People also use Mac for the same purpose too (this process is for ObjC). Installing dependencies on Mac is not as trivial as on Linux.
* I tend to not mess with user's workspace. As you see the process involves building BoringSSL and creating a symbol list file. The BoringSSL in the user's workspace may have their own changes. And I do need to `rm -rf` the artifacts which makes me a bit uncomfortable. Docker just get rid of all those matters.
* We are already using Docker in `tools/distrib/clang_format_code.sh`. 

As to your question, even if we do not use Docker, it would be pretty much combining the 2 shell scripts into one; I don't think there's much to remove there other than the Docker related lines. So I don't think Docker complicate things too.

Reply:
I'd argue most people don't have docker installed on their macs, so arguably this doesn't really make things easier (you'd need to install docker on mac and build the images first which takes time) but it does make all the scripts more complicated.
I'd prefer if we provided a robust script for regenerating the .h file (that assumes golang a cmake are installed, both are easy to install on mac). We can look into using docker in a followup PR and see if it could further simplify things.
Also, for now just supporting regenerating the prefix .h on linux is completely acceptable - later we can make improvements.